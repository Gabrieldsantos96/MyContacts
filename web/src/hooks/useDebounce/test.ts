import { renderHook } from '@testing-library/react'
import { useDebounce } from '.'

vitest.useFakeTimers()

describe('useDebounce', () => {
  test('returns the initial value', () => {
    const { result } = renderHook(() => useDebounce({ value: 'initialValue' }))
    expect(result.current.debouncedValue).toBe('initialValue')
  })
})
