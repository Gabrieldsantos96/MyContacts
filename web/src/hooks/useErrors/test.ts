import { renderHook, act } from '@testing-library/react'
import useErrors from '.'

describe('useErrors', () => {
  test('initializes with no errors', () => {
    const { result } = renderHook(() => useErrors())

    expect(result.current.errors).toHaveLength(0)
    expect(result.current.noErrors).toBe(true)
  })

  test('adds and removes errors', () => {
    const { result } = renderHook(() => useErrors())

    const errorToAdd = { field: 'username', message: 'Username is required' }

    act(() => {
      result.current.setError([errorToAdd], 'username')
    })

    expect(result.current.errors).toEqual([errorToAdd])
    expect(result.current.noErrors).toBe(false)

    act(() => {
      result.current.removeError('username')
    })

    expect(result.current.errors).toHaveLength(0)
    expect(result.current.noErrors).toBe(true)
  })

  test('replaces existing errors for a field', () => {
    const { result } = renderHook(() => useErrors())

    const initialError = { field: 'email', message: 'Email is required' }
    const updatedError = { field: 'email', message: 'Invalid email format' }

    act(() => {
      result.current.setError([initialError], 'email')
    })

    expect(result.current.errors).toEqual([initialError])

    act(() => {
      result.current.setError([updatedError], 'email')
    })

    expect(result.current.errors).toEqual([updatedError])
  })

  test('gets error message by field name', () => {
    const { result } = renderHook(() => useErrors())

    const error = {
      field: 'password',
      message: 'Password must be at least 8 characters'
    }

    act(() => {
      result.current.setError([error], 'password')
    })

    const errorMessage = result.current.getErrorMessageByFieldName('password')

    expect(errorMessage).toBe(error.message)
  })
})
