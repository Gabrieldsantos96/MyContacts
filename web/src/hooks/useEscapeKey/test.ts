import { renderHook, act } from '@testing-library/react'
import { useEscapeKey } from '.'
import React from 'react'

describe('useModal', () => {
  it('should call when press escape', () => {
    const onCancelMock = vitest.fn()
    renderHook(() => useEscapeKey(onCancelMock))

    act(() => {
      const escapeEvent = new KeyboardEvent('keydown', { key: 'Escape' })
      document.body.dispatchEvent(escapeEvent)
    })

    expect(onCancelMock).toHaveBeenCalled()
  })

  it('should not call for other keys', () => {
    const onCancelMock = vitest.fn()

    renderHook(() => useEscapeKey(onCancelMock))

    act(() => {
      const otherKeyEvent = new KeyboardEvent('keydown', { key: 'Enter' })
      document.body.dispatchEvent(otherKeyEvent)
    })

    expect(onCancelMock).not.toHaveBeenCalled()
  })

  it('Should run useEffect cleanUp return function', () => {
    let isExecutedAfterUnmount = false

    vitest.spyOn(React, 'useEffect').mockImplementationOnce(() => {
      isExecutedAfterUnmount = true
    })

    const { unmount } = renderHook(() => useEscapeKey(vitest.fn()))

    unmount()

    expect(isExecutedAfterUnmount).toBe(true)
  })
})
