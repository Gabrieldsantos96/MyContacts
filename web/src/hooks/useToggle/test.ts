import { renderHook, act } from '@testing-library/react'
import { useToggle } from '.'

describe('useToggle', () => {
  test('initializes with default value', () => {
    const { result } = renderHook(() => useToggle(true))

    expect(result.current.value).toBe(true)
  })

  test('initializes with default value as false when not provided', () => {
    const { result } = renderHook(() => useToggle())

    expect(result.current.value).toBe(false)
  })

  test('toggles the value', () => {
    const { result } = renderHook(() => useToggle(true))

    act(() => {
      result.current.toggle()
    })

    expect(result.current.value).toBe(false)
  })

  test('sets the value directly', () => {
    const { result } = renderHook(() => useToggle(true))

    act(() => {
      result.current.setValue(false)
    })

    expect(result.current.value).toBe(false)
  })
})
