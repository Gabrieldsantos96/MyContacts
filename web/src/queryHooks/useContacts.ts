/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  useQuery,
  useMutation,
  UseQueryResult,
  UseMutationOptions,
  UseMutationResult
} from 'react-query'
import { queryClient } from '../lib/queryClient'
import { showToast } from '@utils/helpers/toastEventManager'
import ContactsService from '../services/ContactsService'
import { ContactEntity } from '@interfaces/Contact'
import delay from '@utils/helpers/delay'

const CONTACTS_KEY = 'contacts'

async function listContacts(filters: string) {
  await delay(1000)
  const response = await ContactsService.listContacts(filters)
  return response.data
}

export function useContacts(filters: string): UseQueryResult<ContactEntity[]> {
  return useQuery({
    queryKey: [CONTACTS_KEY, filters],
    queryFn: () => listContacts(filters),
    retry: false,
    refetchOnWindowFocus: false,
    onError: () =>
      showToast({
        type: 'danger',
        text: 'Erro de servidor interno!'
      })
  })
}

type useDeleteParams = {
  id: string
}

async function deleteContact(id: string) {
  await delay(1000)
  const response = await ContactsService.deleteContact(id)
  return response.data
}

export function useDeleteContact(
  options?: UseMutationOptions<unknown, unknown, useDeleteParams>
): UseMutationResult<unknown, unknown, useDeleteParams> {
  return useMutation<unknown, unknown, useDeleteParams>({
    mutationKey: [CONTACTS_KEY],
    mutationFn: async (params: { id: string }) =>
      await deleteContact(params.id),
    retry: 0,
    onSuccess: () => {
      queryClient.invalidateQueries([CONTACTS_KEY])
      showToast({
        type: 'success',
        text: 'Contato foi deletado com sucesso!'
      })
    },
    onError: () =>
      showToast({
        type: 'danger',
        text: 'Não foi possível excluir o contato!'
      })
  })
}
