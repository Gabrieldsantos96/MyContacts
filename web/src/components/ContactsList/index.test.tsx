import { act, render, screen, userEvent, waitFor } from '@utils/test/test-ui'
import ContactList from '.'
import { contactsMock } from '@test/mocks/contacts'
import * as UseContactsModule from '@queryHooks/useContacts'

vitest.mock('./skeleton.tsx', () => ({
  __esModule: true,
  default: function Mock() {
    return <div data-testid="skeleton" />
  }
}))

vitest.mock('react-router-dom', () => ({
  Link: ({ to, children }) => (
    <a data-testid="link" href={to}>
      {children}
    </a>
  )
}))

describe('<ContactList/>', () => {
  it('should render skeleton if isLoading is true', async () => {
    const { container } = render(
      <ContactList
        orderBy="asc"
        toggleOrderBy={vitest.fn()}
        searchTerm=""
        refetchContacts={vitest.fn()}
        filteredContacts={[]}
        isLoading={true}
        isError={false}
        noResultsForSearch={false}
        dataIsEmpty={false}
        contactBeingDelete={null}
        setContactBeingDelete={vitest.fn()}
      />
    )

    expect(screen.getByTestId('skeleton')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render ContactList with no data', async () => {
    render(
      <ContactList
        orderBy="asc"
        toggleOrderBy={vitest.fn()}
        searchTerm=""
        refetchContacts={vitest.fn()}
        filteredContacts={[]}
        isLoading={false}
        isError={false}
        noResultsForSearch={false}
        dataIsEmpty={true}
        contactBeingDelete={null}
        setContactBeingDelete={vitest.fn()}
      />
    )

    const element = screen.getByText(
      /Você ainda não tem nenhum contato cadastrado! Clique no botão/i
    )

    expect(element).toBeInTheDocument()
  })

  it('should render ContactList with no results for search', async () => {
    render(
      <ContactList
        orderBy="asc"
        toggleOrderBy={vitest.fn()}
        searchTerm=""
        refetchContacts={vitest.fn()}
        filteredContacts={[]}
        isLoading={false}
        isError={false}
        noResultsForSearch={true}
        dataIsEmpty={false}
        contactBeingDelete={null}
        setContactBeingDelete={vitest.fn()}
      />
    )
    const element = screen.getByText(/Nenhum resultado foi encontrado para/)
    expect(element).toBeInTheDocument()
  })

  it('should render ContactList with error', async () => {
    render(
      <ContactList
        orderBy="asc"
        toggleOrderBy={vitest.fn()}
        searchTerm=""
        refetchContacts={vitest.fn()}
        filteredContacts={[]}
        isLoading={true}
        isError={true}
        noResultsForSearch={false}
        dataIsEmpty={false}
        contactBeingDelete={null}
        setContactBeingDelete={vitest.fn()}
      />
    )

    const element = screen.getByText(
      /Ocorreu um erro ao obter os seus contatos!/i
    )
    expect(element).toBeInTheDocument()
  })

  it('should render ContactList with filtered contacts', async () => {
    render(
      <ContactList
        orderBy="asc"
        toggleOrderBy={vitest.fn()}
        searchTerm=""
        refetchContacts={vitest.fn()}
        filteredContacts={contactsMock}
        isLoading={false}
        isError={false}
        noResultsForSearch={false}
        dataIsEmpty={false}
        contactBeingDelete={null}
        setContactBeingDelete={vitest.fn()}
      />
    )
    const contact1 = screen.getByText(/Nome do Contato 1/i)
    const contact2 = screen.getByText(/Nome do Contato 2/i)

    expect(contact1).toBeInTheDocument()
    expect(contact2).toBeInTheDocument()
  })

  it('should render ContactList and call delete when click on trash icon and confirm on modal', async () => {
    const useDeleteContactMock = vitest.spyOn(
      UseContactsModule,
      'useDeleteContact'
    )

    useDeleteContactMock.mockReturnValue({
      mutateAsync: vitest.fn(),
      isLoading: false
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any)

    const setContactBeingDelete = vitest.fn()
    const { getAllByRole } = render(
      <ContactList
        orderBy="desc"
        toggleOrderBy={vitest.fn()}
        searchTerm=""
        refetchContacts={vitest.fn()}
        filteredContacts={contactsMock}
        isLoading={false}
        isError={false}
        noResultsForSearch={false}
        dataIsEmpty={false}
        contactBeingDelete={contactsMock[0]}
        setContactBeingDelete={setContactBeingDelete}
      />
    )
    const deleteButtons = getAllByRole('button')

    act(() => {
      userEvent.click(deleteButtons[1])
    })

    await waitFor(() => {
      expect(setContactBeingDelete).toHaveBeenCalled()
      expect(setContactBeingDelete).toHaveBeenCalledWith(contactsMock[0])
    })

    const deleteModalButton = screen.getByRole('button', { name: /deletar/i })

    act(() => {
      userEvent.click(deleteModalButton)
    })

    await waitFor(() => {
      expect(
        useDeleteContactMock.mock.results[0].value.mutateAsync
      ).toHaveBeenCalled()
      expect(
        useDeleteContactMock.mock.results[0].value.mutateAsync
      ).toHaveBeenCalledWith({ id: contactsMock[0].id })
    })
  })
})
