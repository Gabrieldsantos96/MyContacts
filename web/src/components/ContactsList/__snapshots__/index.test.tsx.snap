// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ContactList/> > should render skeleton if isLoading is true 1`] = `
<header
  class="sc-fsYfdN emxBlT"
>
  <strong>
    0
     contato(s)
  </strong>
  <a
    class="sc-beySPh iVMLCk"
    href="/new"
  >
    <span>
      Novo contato
    </span>
  </a>
</header>
`;
